source/Octopus.Tentacle.Tests/Client/RpcCallRetryHandlerFixture.cs

```diff
@@ -0,0 +1,576 @@
+                    async ct =>
+                    {
+                        await Task.CompletedTask;
+                        callCount++;
+                        throw new Exception("An error has occurred.");
+                    },
+                    onRetryAction: null,
+                    onTimeoutAction: null,
+                    CancellationToken.None);
+            }
+            catch (Exception ex) when (ex.Message == "An error has occurred.")
+            {
+
+            }
+
+            callCount.Should().Be(1);
+        }
+
+
+        [TestCase(TimeoutStrategy.Optimistic)]
+        [TestCase(TimeoutStrategy.Pessimistic)]
+        public async Task DoesNotRetryIfNoExceptionOccurs(TimeoutStrategy timeoutStrategy)
+        {
+            var callCount = 0;
+
+            var handler = new RpcCallRetryHandler(TimeSpan.FromSeconds(60), timeoutStrategy);
+
+            await handler.ExecuteWithRetries(
+                async _ =>
+                {
+                    await Task.CompletedTask;
+                    callCount++;
+
+                    return Guid.NewGuid();
+                },
+                onRetryAction: null,
+                onTimeoutAction: null,
+                CancellationToken.None);
+
+            callCount.Should().Be(1);
+        }
+
+
+        [TestCase(TimeoutStrategy.Optimistic)]
+        [TestCase(TimeoutStrategy.Pessimistic)]
+        public async Task RetriesExceptionsForTheConfiguredTimeoutBeforeCancelling(TimeoutStrategy timeoutStrategy)
+        {
+            var callCount = 0;
+
+            var handler = new RpcCallRetryHandler(TimeSpan.FromSeconds(10), timeoutStrategy);
+
+            var stopWatch = new Stopwatch();
+
+            try
+            {
+                await handler.ExecuteWithRetries<Guid>(
+                    async ct =>
+                    {
+                        if (!stopWatch.IsRunning)
+                        {
+                            stopWatch.Start();
+                        }
+                        await Task.CompletedTask;
+                        callCount++;
+                        throw new HalibutClientException("An error has occurred.");
+                    },
+                    onRetryAction: null,
+                    onTimeoutAction: async (_, _) =>
+                    {
+                        await Task.CompletedTask;
+                        stopWatch.Stop();
+                    },
+                    CancellationToken.None);
+            }
+            catch (HalibutClientException) { }
+
+            stopWatch.Elapsed.Should().BeGreaterOrEqualTo(TimeSpan.FromSeconds(9)).And.BeLessThan(TimeSpan.FromSeconds(20));
+            callCount.Should().BeGreaterThan(1);
+        }
+
+
+        [TestCase(TimeoutStrategy.Optimistic)]
+        [TestCase(TimeoutStrategy.Pessimistic)]
+        public async Task CancelsTheExecutingActionIfItIsARetryAfterTheTimeout(TimeoutStrategy timeoutStrategy)
```

Comment:
This failed here https://build.octopushq.com/buildConfiguration/TeamFireAndMotion_OctopusTentacleNet48_SensibleDefaultsUnitTestLinux/7821407?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandPull+Request+Details=true&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true